src/openboardview/FileFormats/GenCADFile.cpp

```diff
@@ -410,88 +434,43 @@ bool GenCADFile::parse_board_outline(mpc_ast_t *board_ast) {
-			if (!rectangle) continue;
-			mpc_ast_t *x_ast = mpc_ast_get_child(rectangle, "x|number|regex");
-			mpc_ast_t *y_ast = mpc_ast_get_child(rectangle, "y|number|regex");
-			mpc_ast_t *w_ast = mpc_ast_get_child(rectangle, "width|number|regex");
-			mpc_ast_t *h_ast = mpc_ast_get_child(rectangle, "height|number|regex");
-			if (!x_ast || !y_ast || !w_ast || !h_ast) continue;
-
-			BRDPoint p;
-
-			p.x = board_unit_to_brd_coordinate(atof(x_ast->contents));
-			p.y = board_unit_to_brd_coordinate(atof(y_ast->contents));
-			format.push_back(p);
-
-			p.y += board_unit_to_brd_coordinate(atof(h_ast->contents));
-			format.push_back(p);
-
-			p.x += board_unit_to_brd_coordinate(atof(w_ast->contents));
-			format.push_back(p);
-
-			p.y -= board_unit_to_brd_coordinate(atof(h_ast->contents));
-			format.push_back(p);
-
-			num_format += 4;
-			i++;
-		}
-	}
 	return true;
 }
 
-void GenCADFile::add_arc_to_outline(mpc_ast_t *start, mpc_ast_t *stop, mpc_ast_t *center) {
-	BRDPoint p1{}, p2{}, pc{};
-	if (!x_y_ref_to_brd_point(start, &p1)) return;
-
-	if (!x_y_ref_to_brd_point(stop, &p2)) return;
+std::vector<std::pair<BRDPoint, BRDPoint>> GenCADFile::arc_to_segments(mpc_ast_t *start, mpc_ast_t *stop, mpc_ast_t *center) {
+	std::vector<std::pair<BRDPoint, BRDPoint>> arc_segments{};
 
-	if (!x_y_ref_to_brd_point(center, &pc)) return;
+	BRDPoint p1{}, p2{}, pc{};
+	if (!x_y_ref_to_brd_point(start, &p1)) return arc_segments;
 
-	double r     = distance(p1, pc);
-	double delta = distance(p1, p2);
+	if (!x_y_ref_to_brd_point(stop, &p2)) return arc_segments;
 
-	double startAngle = asin((p1.y - pc.y) / r);
-	if (p1.y == pc.y && p1.x < pc.x) startAngle = M_PI;
+	if (!x_y_ref_to_brd_point(center, &pc)) return arc_segments;
 
-	double endAngle = startAngle + (asin((delta / 2.0) / r) * 2);
+	double r = distance(p1, pc);
 
-	bool inverseOrder = false;
-	if (format.size()) {
-		double p1toLast = distance(p1, format.back());
-		double p2toLast = distance(p2, format.back());
-		if (p2toLast < p1toLast) inverseOrder = true;
-	}
+	double startAngle = atan2(p1.y - pc.y, p1.x - pc.x);
+	double endAngle = atan2(p2.y - pc.y, p2.x - pc.x);
+	if (endAngle < startAngle) endAngle += 2.0 * M_PI;
 
-	if (!inverseOrder) {
-		for (double i = startAngle; i < endAngle; i += arc_slice_angle_rad) {
-			BRDPoint p{};
-			p.x = pc.x + r * cos(i);
-			p.y = pc.y + r * sin(i);
-			format.push_back(p);
-			num_format++;
+	BRDPoint p{};
+	std::optional<BRDPoint> pold{};
+	for (double i = startAngle; i < endAngle; i += arc_slice_angle_rad) {
+		p.x = pc.x + r * cos(i);
+		p.y = pc.y + r * sin(i);
+		if (pold) {
+				arc_segments.push_back({*pold, p});
 		}
-		format.push_back(p2);
-		num_format++;
-	} else {
-		for (double i = endAngle; i > startAngle; i -= arc_slice_angle_rad) {
-			BRDPoint p{};
-			p.x = pc.x + r * cos(i);
-			p.y = pc.y + r * sin(i);
-			format.push_back(p);
-			num_format++;
-		}
-		format.push_back(p1);
-		num_format++;
+		pold = p;
 	}
+	arc_segments.push_back({p, p2});
```

Comment:
This logic may lead to "artifact" segment from (0;0) if some board would contain arc with p1==p2, since in this case startAngle would be exactly equal to endAngle and for cycle would not run at all.

One of the "simple fix" variants may be initializing p with p1.